panic `findNewLocalEntrypoint called on an empty hnsw graph` with deletion concurrency > 1

### How to reproduce this bug?

1. Start Server with `TOMBSTONE_DELETION_CONCURRENCY=100`
2. Run the script below

```python3
import weaviate
import weaviate.classes as wvc
import numpy as np
import time

client = weaviate.connect_to_local()
client.collections.delete_all()

col = client.collections.create(
    "MyCol",
    vector_index_config=wvc.config.Configure.VectorIndex.hnsw(
        cleanup_interval_seconds=60
    ),
)

# col = client.collections.get("MyCol")

start = time.time()
with col.batch.fixed_size(100, 10) as batch:
    for j in range(50_000):
        batch.add_object(
            {"always_true": True, "i": j},
            vector=np.random.rand(1, 1536)[0].tolist(),
        )
took = time.time() - start
print(f"batch imported in {took:.2f}s")

while True:
    try:
        start = time.time()
        res = col.data.delete_many(
            where=wvc.query.Filter.by_property("i").greater_than(10)
        )
        took = time.time() - start
        print(f"batch deleted in {took:.2f}s")
        if res.matches == 0:
            break
    except Exception as e:
        print(e)
```

### What is the expected behavior?

No panic or crash

### What is the actual behavior?

```
panic: class MyCol: shard N4l1r6BcL6ww: findNewLocalEntrypoint called on an empty hnsw graph

goroutine 179345 [running]:
github.com/weaviate/weaviate/adapters/repos/db/vector/hnsw.(*hnsw).findNewLocalEntrypoint(0x140027c6000, {0x1024f4a40, 0x140101ef358}, 0x0, 0x2)
        /Users/etiennedilocker/code/github.com/weaviate/weaviate/adapters/repos/db/vector/hnsw/delete.go:608 +0x35c
github.com/weaviate/weaviate/adapters/repos/db/vector/hnsw.(*neighborFinderConnector).pickEntrypoint(0x14038fddca8)
        /Users/etiennedilocker/code/github.com/weaviate/weaviate/adapters/repos/db/vector/hnsw/neighbor_connections.go:416 +0xe0
github.com/weaviate/weaviate/adapters/repos/db/vector/hnsw.(*neighborFinderConnector).doAtLevel(0x14038fddca8, 0x0)
        /Users/etiennedilocker/code/github.com/weaviate/weaviate/adapters/repos/db/vector/hnsw/neighbor_connections.go:206 +0x9d8
github.com/weaviate/weaviate/adapters/repos/db/vector/hnsw.(*neighborFinderConnector).Do(0x14038fddca8)
        /Users/etiennedilocker/code/github.com/weaviate/weaviate/adapters/repos/db/vector/hnsw/neighbor_connections.go:81 +0x4c
github.com/weaviate/weaviate/adapters/repos/db/vector/hnsw.(*hnsw).reconnectNeighboursOf(0x140027c6000?, 0x1024e9858?, 0x0?, {0x1401aac5800?, 0x1401aac5800?, 0x600?}, {0x0?, 0x0?}, 0x0?, 0x2?, ...)
        /Users/etiennedilocker/code/github.com/weaviate/weaviate/adapters/repos/db/vector/hnsw/neighbor_connections.go:43 +0x9c
github.com/weaviate/weaviate/adapters/repos/db/vector/hnsw.(*hnsw).reassignNeighbor(0x140027c6000, 0x13, {0x1024f4a40, 0x140101ef350}, 0x1401f82cdd8?)
        /Users/etiennedilocker/code/github.com/weaviate/weaviate/adapters/repos/db/vector/hnsw/delete.go:446 +0x770
github.com/weaviate/weaviate/adapters/repos/db/vector/hnsw.(*hnsw).reassignNeighborsOf.func1()
        /Users/etiennedilocker/code/github.com/weaviate/weaviate/adapters/repos/db/vector/hnsw/delete.go:332 +0x134
golang.org/x/sync/errgroup.(*Group).Go.func1()
        /Users/etiennedilocker/code/github.com/weaviate/weaviate/vendor/golang.org/x/sync/errgroup/errgroup.go:78 +0x58
created by golang.org/x/sync/errgroup.(*Group).Go in goroutine 179339
        /Users/etiennedilocker/code/github.com/weaviate/weaviate/vendor/golang.org/x/sync/errgroup/errgroup.go:75 +0x98
exit status 2

```

### Supporting information

- Concurrent delete cleanup was newly introduced in `v1.24`
- It seems this issue can be mitigated by explicitly setting `TOMBSTONE_DELETION_CONCURRENCY=1` which leads to similar behavior as `v1.23`

### Server Version

`v1.24.1` - seems to have been newly added with v1.24

### Code of Conduct

- [X] I have read and agree to the Weaviate's [Contributor Guide](https://weaviate.io/developers/contributor-guide) and [Code of Conduct](https://weaviate.io/service/code-of-conduct)

--- a/adapters/repos/db/vector/hnsw/delete.go
+++ b/adapters/repos/db/vector/hnsw/delete.go
@@ -243,12 +243,7 @@ func (h *hnsw) cleanUpTombstonedNodes(shouldAbort cyclemanager.ShouldAbortCallba
 	} else if !ok {
 		return executed, nil
 	}
-
-	h.RLock()
-	if deleteList.Contains(h.entryPointID) {
-		h.reassignNeighbor(h.entryPointID, deleteList, breakCleanUpTombstonedNodes)
-	}
-	h.RUnlock()
+	h.reassignNeighbor(h.entryPointID, deleteList, breakCleanUpTombstonedNodes)
 
 	if ok, err := h.replaceDeletedEntrypoint(deleteList, breakCleanUpTombstonedNodes); err != nil {
 		return executed, err

--- a/adapters/repos/db/vector/hnsw/delete.go
+++ b/adapters/repos/db/vector/hnsw/delete.go
@@ -244,6 +244,12 @@ func (h *hnsw) cleanUpTombstonedNodes(shouldAbort cyclemanager.ShouldAbortCallba
 		return executed, nil
 	}
 
+	h.RLock()
+	if deleteList.Contains(h.entryPointID) {
+		h.reassignNeighbor(h.entryPointID, deleteList, breakCleanUpTombstonedNodes)
+	}
+	h.RUnlock()
+
 	if ok, err := h.replaceDeletedEntrypoint(deleteList, breakCleanUpTombstonedNodes); err != nil {
 		return executed, err
 	} else if !ok {
@@ -329,7 +335,9 @@ func (h *hnsw) reassignNeighborsOf(deleteList helpers.AllowList, breakCleanUpTom
 						continue
 					}
 					h.shardedNodeLocks.RUnlock(deletedID)
-					h.reassignNeighbor(deletedID, deleteList, breakCleanUpTombstonedNodes)
+					if h.getEntrypoint() != deletedID {
+						h.reassignNeighbor(deletedID, deleteList, breakCleanUpTombstonedNodes)
+					}
 				}
 			}
 		})

--- a/adapters/repos/db/vector/hnsw/delete.go
+++ b/adapters/repos/db/vector/hnsw/delete.go
@@ -243,7 +243,7 @@ func (h *hnsw) cleanUpTombstonedNodes(shouldAbort cyclemanager.ShouldAbortCallba
 	} else if !ok {
 		return executed, nil
 	}
-	h.reassignNeighbor(h.entryPointID, deleteList, breakCleanUpTombstonedNodes)
+	h.reassignNeighbor(h.getEntrypoint(), deleteList, breakCleanUpTombstonedNodes)
 
 	if ok, err := h.replaceDeletedEntrypoint(deleteList, breakCleanUpTombstonedNodes); err != nil {
 		return executed, err

--- a/adapters/repos/db/vector/hnsw/delete_test.go
+++ b/adapters/repos/db/vector/hnsw/delete_test.go
@@ -1527,6 +1527,8 @@ func TestDelete_WithCleaningUpTombstonesWithHighConcurrency(t *testing.T) {
 		}
 	})
 
+	fmt.Printf("entrypoint after %d\n", vectorIndex.entryPointID)
+
 	t.Run("running the cleanup", func(t *testing.T) {
 		err := vectorIndex.CleanUpTombstonedNodes(neverStop)
 		require.Nil(t, err)

--- a/adapters/repos/db/vector/hnsw/delete_test.go
+++ b/adapters/repos/db/vector/hnsw/delete_test.go
@@ -14,6 +14,7 @@ package hnsw
 import (
 	"context"
 	"fmt"
+	"os"
 	"sort"
 	"sync"
 	"testing"
@@ -1473,4 +1474,69 @@ func TestDelete_WithCleaningUpTombstonesOnceRemovesAllRelatedConnections(t *test
 
 	require.Nil(t, vectorIndex.Drop(context.Background()))
 	store.Shutdown(context.Background())
+}
+
+func TestDelete_WithCleaningUpTombstonesWithHighConcurrency(t *testing.T) {
+	os.Setenv("TOMBSTONE_DELETION_CONCURRENCY", "100")
+	defer os.Unsetenv("TOMBSTONE_DELETION_CONCURRENCY")
+	// there is a single bulk clean event after all the deletes
+	vectors, _ := testinghelpers.RandomVecs(3_000, 1, 1536)
+	var vectorIndex *hnsw
+
+	store := testinghelpers.NewDummyStore(t)
+	defer store.Shutdown(context.Background())
+
+	t.Run("import the test vectors", func(t *testing.T) {
+		index, err := New(Config{
+			RootPath:              "doesnt-matter-as-committlogger-is-mocked-out",
+			ID:                    "delete-test",
+			MakeCommitLoggerThunk: MakeNoopCommitLogger,
+			DistanceProvider:      distancer.NewCosineDistanceProvider(),
+			VectorForIDThunk: func(ctx context.Context, id uint64) ([]float32, error) {
+				return vectors[int(id)], nil
+			},
+			TempVectorForIDThunk: TempVectorForIDThunk(vectors),
+		}, ent.UserConfig{
+			MaxConnections: 30,
+			EFConstruction: 128,
+
+			// The actual size does not matter for this test, but if it defaults to
+			// zero it will constantly think it's full and needs to be deleted - even
+			// after just being deleted, so make sure to use a positive number here.
+			VectorCacheMaxObjects: 100000,
+		}, cyclemanager.NewCallbackGroupNoop(), cyclemanager.NewCallbackGroupNoop(),
+			cyclemanager.NewCallbackGroupNoop(), store)
+		require.Nil(t, err)
+		vectorIndex = index
+
+		for i, vec := range vectors {
+			err := vectorIndex.Add(uint64(i), vec)
+			require.Nil(t, err)
+		}
+	})
+
+	fmt.Printf("entrypoint before %d\n", vectorIndex.entryPointID)
+	t.Run("deleting elements", func(t *testing.T) {
+		for i := range vectors {
+			if i < 10 {
+				continue
+			}
+
+			err := vectorIndex.Delete(uint64(i))
+			require.Nil(t, err)
+		}
+	})
+
+	t.Run("running the cleanup", func(t *testing.T) {
+		err := vectorIndex.CleanUpTombstonedNodes(neverStop)
+		require.Nil(t, err)
+	})
+
+	t.Run("verify the graph no longer has any tombstones", func(t *testing.T) {
+		assert.Len(t, vectorIndex.tombstones, 0)
+	})
+
+	t.Run("destroy the index", func(t *testing.T) {
+		require.Nil(t, vectorIndex.Drop(context.Background()))
+	})
 }

